### DISCUSSION TITLE
Browser Sourcs vs. Window Capture

### DISCUSSION THREAD

**Discussion by Ponjimon** [Rexios80/hds_overlay#55] [General]
Are there any benefits as to why in version 10 the option to have the overlay via browser source has been removed in favor of capturing it via Window capture?

I see little value in this, in fact, it makes it unnecessarily complicated to use becuse you have to adjust the crop of the capture according to your overlay and you have to use a Chroma Key which reduces the overlay's quality drastically and needs tedious fine tuning to look right. Compared to browser source where you just need to add a URL and it displays perfectly fine.

**Comment by Rexios80** [comment]
The value is that users that don't know CSS or want to deal with config files can now do all of that without needing to know how to do it. All the customization options are right there. You can keep using the old version if you like, but I can't guarantee features will keep working with the watch app. Heart rate ranges for example are handled by the new overlay so they are removed from an upcoming watch version. Yes it's more annoying to set up the capture in OBS, but configuration is way better and there are way more customization options and data types to show since it's easier for me to add features. I'm not a web developer. The web app was only created because I didn't have a better solution at the time. It was never meant to be the final version. Also the chroma key quality seems good enough for me. Yeah it's not perfect, but other overlays do the same thing. With some tinkering with the background colors in the app and chroma key settings in OBS I'm sure you could get it even better.

<img width="1199" alt="Screen Shot 2021-04-30 at 2 08 47 PM" src="https://user-images.githubusercontent.com/7896519/116736225-9940b580-a9bd-11eb-9c3a-6b5ad9d75347.png">

**Reply by Ponjimon** [reply]
I see. Would it be possible to support both ways if someone would provide a PR to implement it?

**Reply by Rexios80** [reply]
Implement it into the new version? I don't think it's possible. The app is written in Flutter. I tried looking into hosting web content from a flutter app and hit a dead end.

**Reply by Rexios80** [reply]
You can write web apps in Flutter but you cannot host web pages from a Flutter app. Also Flutter web can't run a websocket server which is needed for communication with the watch since browsers don't allow it. The Apple Watch also can't host a socket server.

**Reply by Ponjimon** [reply]
I don't understand. The browser source part of it wouldn't have to run a server, it's the client. It would only need to talk to the desktop server app and that should definitely be possible. Well, it's open source anywy, I can give it a try :D

**Reply by Rexios80** [reply]
Good luck. I probably won't merge it in unless you somehow get the browser source to show exactly what the Flutter app shows.

**Reply by Rexios80** [reply]
Also it kind of sounds like you're doing something I kind of planned on adding in the future: The ability to send data to other people's apps so they can see multiple sources of data on one overlay. This was possible in the old overlay with a VPN but I want it as a native feature.

**Reply by Rexios80** [reply]
External app would just need to identify themselves to the server with an initial message. Then the Flutter app can handle the rest.

**Reply by Rexios80** [reply]
So maybe don't touch the Flutter app. I have a plan.

**Reply by Ponjimon** [reply]
Ah, so if someone would want to have a browser source, they could just program it themselves and let it talk to the Flutter app? That's indeed exactly what I would want :)

**Reply by Rexios80** [reply]
Exactly. I didn't plan on implementing it for a bit but it shouldn't be hard and it doesn't even need UI. If a socket client identifies itself as an external app then all the Flutter app has to do is route the watch messages to them.

**Reply by Ponjimon** [reply]
Nice! Looking forward to try it out, if it's already possible.

**Reply by Rexios80** [reply]
Actually I just realized I can have clients identify themselves with a name and then add a widget option for the name the data should come from. It would take some refactoring to get it working but it might not be as bad as I first thought.

**Reply by Rexios80** [reply]
But that doesn't affect what you're doing

**Reply by Rexios80** [reply]
By the way here's a link to the Trello card for this if you want to keep track of my progress: https://trello.com/c/cPmc9b2i

**Comment by Rexios80** [comment]
```
  void _handleMessage(WebSocket client, dynamic message) {
    final parts = message.split(':');

    if (parts[0] == 'clientName') {
      clients[client] = parts[1];
      return;
    }

    final source = DataSource(clients[client] ?? 'unknown');

    final dataType =
        EnumToString.fromString(DataType.values, parts[0]) ?? DataType.unknown;
    if (dataType != DataType.unknown) {
      _messageStreamController.add(DataMessage(source, dataType, parts[1]));
    } else {
      _messageStreamController
          .add(UnknownDataMessage(source, parts[0], parts[1]));
    }

    // Broadcast to all clients that aren't the watch
    final externalClients =
        clients.entries.toList().where((e) => e.value != DataSource.watch.name);
    externalClients.forEach((e) => e.key.add(message));
  }
```

**Reply by Rexios80** [reply]
Committed to rewrite/flutter

**Reply by Rexios80** [reply]
Let me know if it works I didn't test it

**Comment by Rexios80** [comment]
Hey uhhh so it turns out I can just add a web target to my Flutter project and it mostly just works. I'm going to have to fix some things since you can't run a socket server on web but it's actually very very possible

**Reply by Rexios80** [reply]
But it still needs the desktop app running so it's going to be kind of weird

**Reply by Rexios80** [reply]
![image](https://user-images.githubusercontent.com/7896519/117223714-452e3a80-addc-11eb-8812-2915507fb307.png)

**Reply by Rexios80** [reply]
Flutter really is something isn't it

**Comment by Rexios80** [comment]
https://hds.dev
